rfcs/0013-ai-policies.md

```diff
@@ -0,0 +1,491 @@
+
+
+### `LLMPromptRiskCheckPolicy`
+
+A Kuadrant `LLMPromptRiskCheckPolicy` is a custom resource provided by Kuadrant that targets Gateway API resources (`Gateway` and `HTTPRoute`), enabling users to define and enforce content safety rules with LLM prompts to detect and block sensitive prompts.  Prompt guards can be defined and enforced for both Gateways and individual HTTPRoutes.
+
+### `LLMResponseRiskCheckPolicy`
+
+A Kuadrant `LLMResponseRiskCheckPolicy` is a custom resource provided by Kuadrant that targets Gateway API resources (`Gateway` and `HTTPRoute`), enabling users to define and enforce content safety rules with LLM content completion responses, to detect and block sensitive responses from LLMs. These can be defined and enforced for both Gateways and individual HTTPRoutes.
+
+## Concepts Introduced
+
+- Token-based rate limiting: Controls usage by number of tokens, not just requests. Helps manage LLM costs.
+- Prompt guarding: Filters risky or unwanted prompts before they hit model servers.
+- Completion guarding: Reviews and filters model outputs before they're returned to users.
+
+## Example Use Case
+
+Say you're serving a chat LLM behind a Gateway, which implements an OpenAI-style chat `/completion` API. You want:
+
+- Free-tier users limited to 20k tokens/day.
+- Guardrails to reject prompts asking for harmful, violent or sexual content.
+- Output filtering to block generally harmful completions.
+
+You’d define the following policies and resources:
+
+> Notes (somewhere): OpenAI-style usage metrics in responses required: `"usage":{"prompt_tokens":5,"total_tokens":15,"completion_tokens":10}`
+
+```yaml
+apiVersion: kuadrant.io/v1
+kind: AuthPolicy
+metadata:
+  name: free-users-auth
+spec:
+  targetRef:
+    group: gateway.networking.k8s.io
+    kind: Gateway
+    name: my-llm-gateway
+  rules:
+    authorization:
+      "free-users":
+        opa:
+          rego: |
+            groups := split(object.get(input.auth.identity.metadata.annotations, "kuadrant.io/groups", ""), ",")
+            allow { groups[_] == "free" }
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: api-key-free-user-1
+    app: my-llm
+  annotations:
+    kuadrant.io/groups: free
+    secret.kuadrant.io/user-id: user-1
+stringData:
+  api_key: iamafreeuser
+type: Opaque
+---
+apiVersion: kuadrant.io/v1alpha1
+kind: LLMTokenRateLimitPolicy
+metadata:
+  name: token-limit-free
+spec:
+  targetRef:
+    group: gateway.networking.k8s.io
+    kind: Gateway
+    name: my-llm-gateway
+  limit:
+    rate:
+      limit: 20000
+      window: 1d
+    predicate: 'request.auth.claims["kuadrant.io/groups"].split(",").exists(g, g == "free")' 
+    counter: auth.identity.userid
+---
+apiVersion: kuadrant.io/v1alpha1
+kind: LLMPromptRiskCheckPolicy
```

Comment:
I wonder could these be in the same API such as `LLMContentPolicy` but split underneath by the kuadrant operator as they are very similar? or do you expect the model being used to change based on the policy type?

```
spec:
  model: ....
  llmprompt:
    categories:
       ....
  llmresponse:
    categories:
response:
...
```

Reply:
They certainly could be in the same API.
My sole reason at the time for splitting them was usability.
That is, writing/configuring each part separately in a sizable block.

There may be a more concrete case for keeping them split.
Thoughts on if you want to apply a prompt check to a different group of users than a risk check?
Although not shown in this example, it could be useful to have a predicate for what users the policy applies to (based on auth headers etc..)
If combined into 1 policy, it would mean having multiple predicate fields.
Do we have a precedence for this with existing policies?

Reply:
yes RLP has that

```
  limits:
    "alice-limit":
      rates:

      - limit: 5
        window: 10s
      when:
      - predicate: "auth.identity.userid == 'alice'"
    "bob-limit":
      rates:
      - limit: 2
        window: 10s
      when:
      - predicate: "auth.identity.userid == 'bob'"
   ```

Reply:
so perhaps its a use case like

spec:
  model: ....
  llmprompt:
    - "under18":
       categories:
          ....
       when:
        - predicate: "auth.identity.age < 18"
      - "over18"
       categories:
          ....
       when:
        - predicate: "auth.identity.age >= 18" 
  llmresponse:
   -  "under18":

```

This is just sudo stuff but might be useful to think about?

Reply:
nit: `request.model` seem to be wanting to "enhance" the `request` struct with arbitrary fields. I'd advise against that. When is that `model` field present? I don't want users to ask themselves these questions. I understand this is just for illustration purposes, but nonetheless raises an interesting question: where would "additional policies" append data to the "well-known attributes"?

Reply:
Yes where we append this data is an important consideration. Perhaps we need a new namespace for AI metadata. ```ai.model```